conslogging/bundle.go

```diff
@@ -0,0 +1,246 @@
+const ansi = "[\u001B\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[a-zA-Z\\d]*)*)?\u0007)|(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PRZcf-ntqry=><~]))"
+
+var re = regexp.MustCompile(ansi)
+
+func (bb *BundleBuilder) GetCommandAndSummary(prefix string, builder *strings.Builder) (string, string) {
+	rawText := builder.String()
+	text := re.ReplaceAllString(rawText, "")
+
+	prettyPrefix := prettyPrefix2(DefaultPadding, prefix)
+
+	// regex to find command lines in the output.
+	regexStr := fmt.Sprintf(`(?m)^%s \| (\*cached\* |\*local\* | )*--> `, regexp.QuoteMeta(prettyPrefix))
+	r := regexp.MustCompile(regexStr)
+	matches := r.FindAllStringIndex(text, -1)
+	if len(matches) == 0 {
+		return "", ""
+	}
+
+	// Take the last match, and use the first line up to 120 characters, or first newlines... whichever comes first.
+	lastMatch := matches[len(matches)-1]
+	remainder := text[lastMatch[1]:]          // The rest of the log from end of the last match
+	command := TruncateString(remainder, 120) // The line up to a newline or 120 chars
+
+	// regex to get the last line, (ab)use groups to get the line without the prefix. Truncate it like command.
+	regexStr2 := fmt.Sprintf(`%s \| (.*)\n?$`, regexp.QuoteMeta(prettyPrefix))
+	r2 := regexp.MustCompile(regexStr2)
+	matches2 := r2.FindAllStringSubmatch(remainder, -1)
+	if len(matches2) == 0 {
+		return command, ""
+	}
+	summary := TruncateString(matches2[len(matches2)-1][1], 120)
+
+	return command, summary
+}
+
+// TruncateString truncates a string honoring weird glyphs that are larger than one byte... like Japanese.
+func TruncateString(str string, length int) string {
+	if length <= 0 {
+		return ""
+	}
+
+	truncated := ""
+	count := 0
+	for _, char := range str {
+		if char == '\n' {
+			break
+		}
+		truncated += string(char)
+		count++
+		if count >= length {
+			break
+		}
+	}
+	return truncated
+}
+
+// --- Copied types below
+
+const (
+	StatusWaiting    = "waiting"
+	StatusInProgress = "in_progress"
+	StatusComplete   = "complete"
+	StatusCancelled  = "cancelled"
+
+	ResultSuccess   = "success"
+	ResultFailure   = "failure"
+	ResultCancelled = "cancelled"
+)
+
+type Manifest struct {
+	Version    int              `json:"version"`
+	Duration   int              `json:"duration"`
+	Status     string           `json:"status"`
+	Result     string           `json:"result"`
+	CreatedAt  time.Time        `json:"created_at"`
+	Targets    []TargetManifest `json:"targets"`
+	Entrypoint string           `json:"entrypoint"`
+}
+
+type TargetManifest struct {
+	Name     string `json:"name"`
+	Status   string `json:"status"`
+	Result   string `json:"result"`
+	Duration int    `json:"duration"`
+	Size     int    `json:"size"`
+	Command  string `json:"command,omitempty"`
+	Summary  string `json:"summary,omitempty"`
+}
+
+type Permissions struct {
+	Version int      `json:"version"`
+	Users   []uint64 `json:"users"`
+	Orgs    []uint64 `json:"orgs"`
+}
```

Comment:
How should we share these structs? Perhaps a target to copy them from our server code?

Reply:
Was toying with the idea of a shared proto here.

Reply:
This may still need to be solved before we merge.

Reply:
I broke these out into their own file, with a comment. Once its merged, I'll import this from our server-side code.